### PR TITLE
Luacov

### PR DESCRIPTION
**Pull request by TitanSnow**
#90 

add coverage in CI script

*Not finished yet, do not merge now*

I also have a plan to add C extension's coverage

### CODE REVIEW

**Comment by TitanSnow** [comment]
For some unknown reasons, coverage of C part only works on Mac OS. I have no idea on how to deal with it. Maybe the version of gcc on linux on travis-CI is too low

**Comment by TitanSnow** [comment]
A result could be found [here](https://codecov.io/gh/tboox/xmake/tree/72292a6a32ed578eedbcd6e1f1739fd664663f98). 50%, not so good

**Comment by waruqi** [comment]
This is because we only use some basic APIs in `xmake/tests`  now.

**Comment by TitanSnow** [comment]
Right. So tons of tests are needed

**Comment by TitanSnow** [comment]
Done. It nearly makes me mad that C coverage does not work on linux. I quit. Do you have any idea on it? :disappointed:

**Comment by waruqi** [comment]
Ok, I will look at it.

**Comment by TitanSnow** [comment]
This branch allows you to push. If you found a solution, you could directly push to this branch. It might be clear in history @waruqi

**Comment by waruqi** [comment]
got it

**Comment by waruqi** [comment]
@TitanSnow It works on my ubuntu for your script in .traivis.yml

```bash
... 

File 'src/xmake/machine.c'
Lines executed:83.33% of 102
Branches executed:84.21% of 76
Taken at least once:43.42% of 76
Calls executed:85.25% of 61
Creating 'src#xmake#machine.c.gcov'
Cannot open source file src/xmake/machine.c

Lines executed:48.06% of 21284
==> Python coveragepy not found
==> Searching for coverage reports in:
    + .
    -> Found 117 reports
==> Detecting git/mercurial file structure
==> Reading reports
    + ./src#xmake#os#exists.c.gcov bytes=1583
    + ./src#luajit#src#lj_asm.c.gcov bytes=70959
    + ./src#xmake#os#tmpdir.c.gcov bytes=1547
    + ./src#luajit#src#lj_lex.c.gcov bytes=2

...
```

**Comment by TitanSnow** [comment]
So does my ubuntu. So I'm wondering if the version of gcc is too low on travis-ci

File: core/xmake.lua [hunk 1]
```diff
@@ -50,8 +50,8 @@ end
 
 -- for mode coverage
 if is_mode("coverage") then
-    add_cxflags("-coverage")
-    add_ldflags("-coverage")
+    add_cxflags("-coverage", "-fprofile-arcs", "-ftest-coverage")
+    add_ldflags("-coverage", "-fprofile-arcs", "-ftest-coverage")
```
Comment (TitanSnow) [inline core/xmake.lua]: What does these flags means:question:

Comment (waruqi) [inline core/xmake.lua]: It doesn't seem to be able to generate '*.gcno' and '*.gcda' files for old version of gcc with flags `-coverage`. But `"-fprofile-arcs", "-ftest-coverage"` works.

Comment (waruqi) [inline core/xmake.lua]: You can see [travis-ci](https://travis-ci.org/tboox/xmake/jobs/232260363). 

And the following path rules must be guaranteed for source code and "*.gcno" for old gcc version.

```bash
├── src
│   ├── main.c
│   └── main.gcno
```

Comment (waruqi) [inline core/xmake.lua]: The old version gcov does not support the following structure generated by `xmake run`: 

```
./build/.objs/
└── testgcov
    └── src
        ├── main.c.d
        ├── main.c.gcda
        ├── main.c.gcno
        └── main.c.o
├── src
│   ├── main.c
```

Comment (TitanSnow) [inline core/xmake.lua]: I have a way to upgrade gcov on travis-ci. I'm going to have a try

**Comment by TitanSnow** [comment]
I have upgraded gcov to version 6 on travis-ci but seems has no use. Maybe upgrade gcc is also needed @waruqi

**Comment by waruqi** [comment]
@TitanSnow Right, Incompatible `*.gcno` file version.  `./build/.objs/testgcov/src/main.c.gcno:version '406*', prefer '602*'`

**Comment by TitanSnow** [comment]
I have upgraded all. Maybe works

**Comment by TitanSnow** [comment]
It works! :tada:

**Comment by waruqi** [comment]
Great! 👍 

I think we can remove these flags `"-fprofile-arcs", "-ftest-coverage"` if gcc and gcov have been updated to new version.

File: core/pkg/tbox.pkg/lib/coverage [hunk 2]
```diff
@@ -0,0 +1 @@
+release
```
Comment (waruqi) [inline core/pkg/tbox.pkg/lib/coverage]: What does this file do?

Comment (TitanSnow) [inline core/pkg/tbox.pkg/lib/coverage]: A soft link

**Comment by TitanSnow** [comment]
@waruqi No, we cannot. I have tried but it fails

**Comment by TitanSnow** [comment]
I wanna remove coverage of luajit

**Comment by uael** [comment]
Got same issue on [jayl/jay](https://github.com/jayl/jay), finally get it it work with this configuration !
```yml
matrix:
  include:
    # coverage build
    - os: linux
      dist: trusty
      compiler: gcc-5
      addons:
        apt:
          sources:
            - george-edison55-precise-backports
            - ubuntu-toolchain-r-test
          packages:
            - cmake
            - cmake-data
            - gcc-5
      install:
        - mkdir build && cd build
        - cmake -DCOVERAGE=ON -DCMAKE_BUILD_TYPE=Debug ..
        - make
        - cd ..
      after_success:
        - bash <(curl -s https://codecov.io/bash) -xgcov-5
```
 (george-edison55-precise-backports is only for cmake)

**Comment by waruqi** [comment]
👍

**Comment by waruqi** [comment]
@uael I will try it. Thanks!

**Comment by TitanSnow** [comment]
@uael Great, in this way sudo is not required anymore!

**Comment by TitanSnow** [comment]
@waruqi In fact, codecov's coverage does not count the files not ran into so that it's higher than real number

**Comment by waruqi** [comment]
@uael No homebrew addons on Travis? like: 

```
addons:
        homebrew:
          packages:
            - dmd
            - rust
```

**Comment by uael** [comment]
@waruqi 
[still waiting](https://github.com/travis-ci/travis-build/pull/947) :cry:

**Comment by waruqi** [comment]
@uael 😓

**Comment by TitanSnow** [comment]
As I know `sudo: false` does not influence osx

**Comment by TitanSnow** [comment]
I found `./core/build/.objs/xmake/src/xmake/os/chdir.c.gcno:version '408*', prefer '504*'` in build. What's that?

**Comment by waruqi** [comment]
@TitanSnow I have fixed it.

**Pull request merged**